cql3/statements/select_statement.cc

```diff
@@ -1499,70 +1502,132 @@ void select_statement::verify_ordering_is_allowed(const restrictions::statement_
     }
 }
 
-void select_statement::validate_distinct_selection(const schema& schema,
-                                                   const selection::selection& selection,
-                                                   const restrictions::statement_restrictions& restrictions)
+void select_statement::handle_unrecognized_ordering_column(const column_identifier& column) const
 {
-    if (restrictions.has_non_primary_key_restriction() || restrictions.has_clustering_columns_restriction()) {
-        throw exceptions::invalid_request_exception(
-            "SELECT DISTINCT with WHERE clause only supports restriction by partition key.");
+    if (contains_alias(column)) {
+        throw exceptions::invalid_request_exception(format("Aliases are not allowed in order by clause ('{}')", column));
     }
-    for (auto&& def : selection.get_columns()) {
-        if (!def->is_partition_key() && !def->is_static()) {
-            throw exceptions::invalid_request_exception(format("SELECT DISTINCT queries must only request partition key columns and/or static columns (not {})",
-                def->name_as_text()));
+    throw exceptions::invalid_request_exception(format("Order by on unknown column {}", column));
+}
+
+select_statement::prepared_orderings_type select_statement::prepare_orderings(const schema& schema) const {
+    prepared_orderings_type prepared_orderings;
+    prepared_orderings.reserve(_parameters->orderings().size());
+
+    for (auto&& [column_id, column_ordering] : _parameters->orderings()) {
+        ::shared_ptr<column_identifier> column = column_id->prepare_column_identifier(schema);
+
+        const column_definition* def = schema.get_column_definition(column->name());
+        if (def == nullptr) {
+            handle_unrecognized_ordering_column(*column);
+        }
+
+        if (!def->is_clustering_key()) {
+            throw exceptions::invalid_request_exception(format("Order by is currently only supported on the clustered columns of the PRIMARY KEY, got {}", *column));
         }
+
+        prepared_orderings.emplace_back(def, column_ordering);
     }
 
-    // If it's a key range, we require that all partition key columns are selected so we don't have to bother
-    // with post-query grouping.
-    if (!restrictions.is_key_range()) {
+    // Uncomment this to allow specifing ORDER BY columns in any order.
```

Comment:
Why not add this code as a patch instead of commented-out code?
I think either we want this functionality - in which case we want it as code, not as comment - or if we don't want it, we also don't need it as a comment...

Reply:
I wanted to keep this PR focused on one thing.

Adding support for this is not as simple as uncommenting the code, we would have to modify existing tests and maybe add new ones. I can oppen an issue for this.

Reply:
Well, the comment suggests that uncommenting it might be enough. Anyway, I'm not against this commented-out code but I think it will be even better to create an issue explaining exactly what's missing (is it something that Cassandra has and we're missing, or it also doesn't work in Cassandra) and then refer to that issue number in this comment (and vice versa).